ambry-cloud/src/main/java/com.github.ambry.cloud/CloudBlobStore.java

```diff
@@ -207,8 +207,14 @@ public void updateTtl(MessageWriteSet messageSetToUpdate) throws StoreException
     // Note: we skipped uploading the blob on PUT record if the TTL was below threshold.
     try {
       for (MessageInfo msgInfo : messageSetToUpdate.getMessageSetInfo()) {
-        BlobId blobId = (BlobId) msgInfo.getStoreKey();
-        cloudDestination.updateBlobExpiration(blobId, msgInfo.getExpirationTimeInMs());
+        // MessageInfo.expirationTimeInMs is not reliable if ttlUpdate is set. See {@code PersistentIndex#findKey()}
+        // and {@code PersistentIndex#markAsPermanent()}.
+        if (msgInfo.isTtlUpdated()) {
+          BlobId blobId = (BlobId) msgInfo.getStoreKey();
+          cloudDestination.updateBlobExpiration(blobId, Utils.Infinite_Time);
```

Comment:
I was trying to keep the fact that UpdateTTL makes the blob permanent out of the cloud logic, but that may not be possible if expirationTime is not reliable.

Reply:
Yep. Current ttlUpdate expiration time is not always -1.

Reply:
Maybe just add a comment that if we ever change updateTTL to be more flexible, that this code will need to be modified.